### PR TITLE
WDL Draft 2 workflow source + imports => namespace cache 

### PR DESCRIPTION
**Pull request by mcovarr**
Particularly useful for not re-parsing workflows for every set of inputs in a batch submission.

### CODE REVIEW

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala [hunk 1]
```diff
@@ -132,4 +140,10 @@ object WdlDraft2LanguageFactory {
 
   val httpResolver = resolverConverter(ImportResolver.HttpResolver())
   def httpResolverWithHeaders(headers: Map[String, String]) = resolverConverter(ImportResolver.HttpResolver(headers = headers))
+
+  private val namespaces: Cache[WorkflowSource, ErrorOr[WdlNamespaceWithWorkflow]] = CacheBuilder.newBuilder()
+    .concurrencyLevel(2)
+    .expireAfterWrite(20, TimeUnit.MINUTES)
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: ooh I like that

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala [hunk 2]
```diff
@@ -1,32 +1,36 @@
...
+import wom.core.{WorkflowJson, WorkflowOptionsJson, WorkflowSource}
 import wom.executable.WomBundle
 import wom.expression.IoFunctionSet
-import wom.transforms.WomExecutableMaker.ops._
+import wom.graph.GraphNodePort.OutputPort
 import wom.transforms.WomBundleMaker.ops._
-import wom.values.WomValue
-import languages.wdl.draft2.WdlDraft2LanguageFactory._
+import wom.transforms.WomExecutableMaker.ops._
+import wom.values.{WomValue, _}
```
Comment (kshakir) [inline languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: This import, and others imports this file, have been Intellij-ed and should be fixed.

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala [hunk 3]
```diff
@@ -54,19 +58,23 @@ class WdlDraft2LanguageFactory(override val config: Map[String, Any]) extends La
...
-      case w: WorkflowSourceFilesWithDependenciesZip =>
-        for {
-          importsDir <- LanguageFactoryUtil.validateImportsDirectory(w.importsZip)
-          wf <- WdlNamespaceWithWorkflow.load(workflowSource, importResolvers map resolverConverter).toErrorOr
-          _ = importsDir.delete(swallowIOExceptions = true)
-        } yield wf
-      case _: WorkflowSourceFilesWithoutImports =>
-        WdlNamespaceWithWorkflow.load(workflowSource, importResolvers map resolverConverter).toErrorOr
+    def workflowHashKey: String = {
+      workflowSource.md5Sum + (source.importsZipFileOption map { bytes => new String(MessageDigest.getInstance("MD5").digest(bytes)) }).getOrElse("")
```
Comment (cjllanwarne) [inline languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: I think so but to double check - are we happy to assert that MD5suming the zip file is less CPU intensive than just parsing the WDL?

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala [hunk 4]
```diff
@@ -54,19 +58,23 @@ class WdlDraft2LanguageFactory(override val config: Map[String, Any]) extends La
...
     }
 
+    lazy val wdlNamespaceValidation: ErrorOr[WdlNamespaceWithWorkflow] = namespaces.get(workflowHashKey, new Callable[ErrorOr[WdlNamespaceWithWorkflow]] {
```
Comment (cjllanwarne) [inline languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: The RHS of this is `get` is lazy, right...?

Comment (cjllanwarne) [reply languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: I wonder if we could make that explicit by separating it as 
```scala
def: lazyWorkflowMaker: () => Callable[ErrorOr[WdlNamespaceWithWorkflow]] = ...
lazy val wdlNamespaceValidation: ErrorOr[WdlNamespaceWithWorkflow] = namespaces.get(workflowHashKey, lazyWorkflowMaker)
```

**Comment by cjllanwarne** [comment]
Since this is the `develop` version: If I told you that one day I intend to shuffle this logic into the `LanguageFactory` trait and make it apply equally to WDL 1.0, 1.1, CWL, etc...

1. Would you trust me to do the right thing; _or..._
1. Would you want to add a regression test to make sure I don't ruin this great work?

**Comment by geoffjentry** [comment]
@cjllanwarne You're already on thin ice w/ test recommendations

**Comment by cjllanwarne** [comment]
@geoffjentry alright, alright, no need to get [testy](https://www.merriam-webster.com/dictionary/testy)...

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala (hunk 3 above)
Comment (mcovarr) [reply languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: It seems a pretty safe assumption that digesting the array of bytes is cheaper than stringifying it and parsing. 🙂

File: languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala (hunk 4 above)
Comment (mcovarr) [reply languageFactories/wdl-draft2/src/main/scala/languages/wdl/draft2/WdlDraft2LanguageFactory.scala]: yes this is lazy.

File: cromwell.examples.conf [hunk 5]
```diff
@@ -216,6 +216,16 @@ languages {
         # config {
         #   strict-validation: true
         #   enabled: true
+        #   caching {
+        #     # WDL Draft 2 namespace caching is off by default, this value must be set to true to enable it.
+        #     enabled: false
+        #     # Guava cache concurrency
+        #     concurrency: 2
```
Comment (cjllanwarne) [inline cromwell.examples.conf]: What does concurrency mean in cache terms?

File: cromwell.examples.conf [hunk 6]
```diff
@@ -216,6 +216,16 @@ languages {
...
+        #     # How long entries in the cache should live from the time of their last access.
+        #     ttl: 20 minutes
+        #     # Maximum size of the cache.
```
Comment (cjllanwarne) [inline cromwell.examples.conf]: (in terms of WdlNamespace objects stored)

File: CHANGELOG.md [hunk 7]
```diff
@@ -7,14 +7,22 @@
 Cromwell now allows for a user to submit the URL pointing to workflow file to run a workflow using `workflowUrl` parameter. Currently, this is only supported in `Server` mode.
 More details on how to use it can be found [here](http://cromwell.readthedocs.io/en/develop/api/RESTAPI/).
 
+### Languages
+
+- Added an opt-in namespace cache for the WDL Draft 2 language factory. Please see the Cromwell example configuration for details.
```
Comment (cjllanwarne) [inline CHANGELOG.md]: Maybe warn that if people are upgrading from hotfix, their currently enabled cache is now opt-in rather than on-by-default?

**Pull request merged**